pkg/services/folder/folderimpl/folder_unifiedstorage.go

```diff
@@ -0,0 +1,698 @@
+package folderimpl
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/exp/slices"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+
+	"github.com/grafana/grafana/pkg/apimachinery/identity"
+	"github.com/grafana/grafana/pkg/events"
+	"github.com/grafana/grafana/pkg/infra/metrics"
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+	"github.com/grafana/grafana/pkg/services/apiserver/endpoints/request"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	"github.com/grafana/grafana/pkg/services/dashboards/dashboardaccess"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/folder"
+	"github.com/grafana/grafana/pkg/services/guardian"
+	"github.com/grafana/grafana/pkg/services/store/entity"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util"
+)
+
+// interface to allow for testing
+type folderK8sHandler interface {
+	getClient(ctx context.Context, orgID int64) (dynamic.ResourceInterface, bool)
+	getNamespace(orgID int64) string
+}
+
+var _ folderK8sHandler = (*foldk8sHandler)(nil)
+
+type foldk8sHandler struct {
+	cfg        *setting.Cfg
+	namespacer request.NamespaceMapper
+	gvr        schema.GroupVersionResource
+}
+
+func (s *Service) getFoldersFromApiServer(ctx context.Context, q folder.GetFoldersQuery) ([]*folder.Folder, error) {
+	if q.SignedInUser == nil {
+		return nil, folder.ErrBadRequest.Errorf("missing signed in user")
+	}
+
+	qry := folder.NewGetFoldersQuery(q)
+	permissions := q.SignedInUser.GetPermissions()
+	folderPermissions := permissions[dashboards.ActionFoldersRead]
+	qry.AncestorUIDs = make([]string, 0, len(folderPermissions))
```

Comment:
I think it would be good to get some of the IAM team to review to make sure the auth side is looking okay

Reply:
this part if plain copy&paste from the original `GetFolders`

https://github.com/grafana/grafana/blob/0e953a8ff379364a5c29bb8b760a9985f4e99eee/pkg/services/folder/folderimpl/folder.go#L145

I can ask for a review from the IAM team though. I think it's a good idea.

Reply:
Ah perfect!